ARROW-9944: [Rust][DataFusion] Implement to_timestamp function

alamb opened the pull request:
This PR adds:
1. An implementation of the 1 argument form of the `TO_TIMESTAMP` function that takes a single argument (string) and converts it into a nanosecond precision timestamp, as described in the proposal [doc](https://docs.google.com/document/d/18O9YPRyJ3u7-58J02NtNVYb6TDWBzi3mIQC58VhwxUk/edit#heading=h.2nfzn4ggyl34)
2. An explicit dependence on the `chrono` library. Note that this does not add any *actual* new dependencies for datafusion because depends on `arrow` which already depends on chrono [source link](https://github.com/apache/arrow/blob/master/rust/arrow/Cargo.toml#L49).

fyi @jorgecarleitao @andygrove @emkornfield @jhorstmann

andygrove reviewed (approved):
LGTM!

andygrove commented on rust/datafusion/src/physical_plan/datetime_expressions.rs:
```diff
@@ -0,0 +1,225 @@
+    let string_args = &args[0]
+        .as_any()
+        .downcast_ref::<StringArray>()
+        .expect("input cast to StringArray failed");
```
nit: Although the planner and type coercion logic should prevent errors here, we could actually return an `Err` result rather than panic.

alamb replied:
Will do

jorgecarleitao commented on rust/datafusion/src/physical_plan/datetime_expressions.rs:
```diff
@@ -0,0 +1,225 @@
+    // strings and we don't know which the user was trying to
+    // match. Ths any of the specific error messages is likely to be
+    // be more confusing than helpful
+    Err(ExecutionError::General("Timestamp parse error".into()))
```
add `s` that failed, or just a slice of it, here?

alamb replied:
good idea -- I will do so

jorgecarleitao commented on rust/datafusion/src/physical_plan/datetime_expressions.rs:
```diff
@@ -0,0 +1,225 @@
+        .downcast_ref::<StringArray>()
+        .expect("input cast to StringArray failed");
+
+    for i in 0..string_args.len() {
```
I think that we get a massive performance boost if we build `ArrayData` here (learning from @nevi-me :P). 

@nevi-me did this recently [here](https://github.com/apache/arrow/commit/9ea24092064205b9966c4e08da50ea344bf042e5#diff-084bc9b19a2397f6ba80602e2d136833R33), e.g.

```rust

let num_rows = args[0].len();
let string_args = &args[0]
    .as_any()
    .downcast_ref::<StringArray>()
    .expect("input cast to StringArray failed");

let result = (0..num_rows).map(|i| string_to_timestamp_nanos(string_args.value(i))).collect::<Result<Vec<i64>>>()?;

let data = ArrayData::new(
    DataType::Timestamp(DataType::TimeUnit::Nanosecond, None),
    num_rows,
    Some(string_args.null_count()),
    string_args.data().null_buffer().cloned(),
    0,
    vec![Buffer::from(result.to_byte_slice())],
    vec![],
)
```

jorgecarleitao reviewed (approved):
LGTM.

jhorstmann commented on rust/datafusion/src/physical_plan/datetime_expressions.rs:
```diff
@@ -0,0 +1,225 @@
+use chrono::prelude::*;
+
+#[inline]
+fn string_to_timestamp_nanos(s: &str) -> Result<i64> {
```
I'd be interested in a benchmark of this function or the kernel, maybe one for the happy case with `T` and `Z` and one for the last fallback. From my experience with Java, [parsing timestamps can be rather slow][1] and it might be worth writing a specialized implementation.

The behaviour regarding time zones should also be documented, does it default to a local time zone or to always to UTC?

Lastly, storing nanoseconds as i64 gives a range of about year 1677 to 2262, this might be a limitation for currently very few usecases but should also be mentioned.

 [1]: https://github.com/jhorstmann/packedtime#benchmarks

alamb replied:
I filed https://issues.apache.org/jira/browse/ARROW-9955 to track the (excellent) suggestion for benchmark and possible improvement based on results. I agree efficient timestamp parsing is a bit of an art form. 

I will also document both the timezone behavior as well as the date range limitations

alamb commented:
@andygrove  -- I have rebased this PR and I believe I have implemented all suggestions. There is one more lingering thing regarding parsing local timestamps that I want to check out but I can also do that after merge if you prefer.
